nixos/modules/services/security/vault-agent.nix

```diff
@@ -0,0 +1,145 @@
+{config, lib, pkgs, ...}:
+
+with lib;
+
+let
+  cfg = config.services.vault-agent;
+  createAgentInstance = cfg: name:
+    let
+      configFile = pkgs.writeText "${name}.hcl" ''
+          pid_file = "/run/vault-agent/${name}.pid"
+          vault {
+            address = "${cfg.address}"
+          }
+          ${cfg.extraConfig}
+          ${mkTemplates cfg.templates}
+        '';
+      mkTemplates = templates:
+        concatStringsSep "\n"
+          (mapAttrsToList
+            (template: config: ''
+                template {
+                  source      = "${pkgs.writeText template config.content}"
+                  ${optionalString (config.command != null)
+                  "command     = \"${config.command}\""}
+                  destination = "${config.destination}"
+                  perms       = "${config.perms}"
+                }
+              '')
+              templates);
```

Comment:
I think json should be preferred over hcl here.
It's a more nix-native approach and allows the use of attrsets instead of juggling around with multiline strings.
 
Similar how `services.consul.extraConfig` does it (and hopefully `services.vault` after I am done refactoring it).
This would also open up the possibility to properly implement [RFC42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md) :^)

Reply:
You're totally right ! It would be way better. 
I begun to work on it but again I'm not yet good enough with Nix language part so I don't really see how I could implement the mkTemplates part using Nix syntax. 
Do you have any example that could help me solve that part ?

Reply:
You don't need `mkTemplates` :)

The following hcl config (slightly adapted from the `templates` option's example)
```hcl
template = {
  consul-cert = {
    destination = "/var/lib/consul/tls/cert.pem"
    command = "systemctl restart consul"
    content = <<EOL
{{ with secret "pki/issue/cert" "common_name=node.consul" "ttl=730h" }}
{{ .Data.certificate }}
{{ .Data.issuing_ca }}{{ end }}
EOL
  }
}
```

translates to the following json
```json
{
  "template": [
    {
      "consul-cert": [
        {
          "command": "systemctl restart consul",
          "content": "{{ with secret \"pki/issue/cert\" \"common_name=node.consul\" \"ttl=730h\" }}\n{{ .Data.certificate }}\n{{ .Data.issuing_ca }}{{ end }}\n",
          "destination": "/var/lib/consul/tls/cert.pem"
        }
      ]
    }
  ]
}
```

which can also be written as the following nix attribute set
```nix
{
  template = [
    {
      command = "systemctl restart consul";
      contents = ''
        {{ with secret "pki/issue/cert" "common_name=node.consul" "ttl=730h" }}
        {{ .Data.certificate }}
        {{ .Data.issuing_ca }}{{ end }}
      '';
      destination = "/var/lib/consul/tls/cert.pem";
    }
  ];
}
```

So I imagine something like
```nix
{
  services.vault-agent.instances.example.settings = {
    vault.address = "https://localhost:8200";
    template = [
      {
        command = "systemctl restart consul";
        contents = ''
          {{ with secret "pki/issue/cert" "common_name=node.consul" "ttl=730h" }}
          {{ .Data.certificate }}
          {{ .Data.issuing_ca }}{{ end }}
        '';
        destination = "/var/lib/consul/tls/cert.pem";
      }
    ];
  };
}
``` 

---

You can reach me at https://matrix.to/#/@me:indeednotjames.com (or my mail on my github profile) if you want to work on this together or need some help without going through PR comments back and forth :)
Especially because I don't feel like I did a great job explaining it^^

---

https://www.hcl2json.com/ is pretty handy for this btw